airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/java/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -215,6 +217,45 @@ private String loadDataIntoBucket(final String objectPath, final SerializableBuf
     return newFilename;
   }
 
+  /**
+   * Users want deterministic file names (e.g. the first file part is really foo-0.csv Using UUIDs
+   * (previous approach) doesn't allow that. However, using pure integers could lead to a collision
+   * with an existing file. So, we'll count up the existing files in the directory and use that as a
+   * lazy-offset, assuming airbyte manages the dir and has similar naming conventions. `getPartId`
+   * will be 1-indexed.
+   */
+  @VisibleForTesting
+  String getPartId(String objectPath) {
+    if (partCounts.get(objectPath) == null) {
+      partCounts.put(objectPath, new AtomicInteger(0));
+    }
+
+    final AtomicInteger partCount = partCounts.get(objectPath);
+
+    if (partCount.get() == 0) {
+      final String bucket = s3Config.getBucketName();
+      final ObjectListing objects = s3Client.listObjects(bucket, objectPath);
+
+      // another thread beat us here, so recurse
+      if (partCount.get() != 0) {
+        return getPartId(objectPath);
+      }
+
+      if (objects == null) {
+        partCount.set(0);
+      } else if (objects.isTruncated()) {
+        // The bucket contains too many objects, and there is no reasonable way to get a count of all
+        // objects in a bucket in S3 without multiple requests
+        // So, let's pick a big number and assume overwriting existing files is OK...
+        partCount.set(objects.getObjectSummaries().size() + 101);
+      } else {
+        partCount.set(objects.getObjectSummaries().size());
+      }
+    }
+
+    return Integer.toString(partCount.incrementAndGet());
```

Comment:
```suggestion
    return Integer.toString(partCount.getAndIncrement());
```
judging from [random old slack message](https://airbytehq-team.slack.com/archives/C05DME9BAV8/p1697726669069029?thread_ts=1696956776.207509&cid=C05DME9BAV8) we 0-index

Reply:
Moved to 0-indexing in 3db17a42b0e4b4b6d02bc232b772b110f91333a9